### ISSUE TITLE
Enable configuration of Secure DNS resolver for outgoing name resolution

### ISSUE THREAD

**Issue by fpietrosanti** [globaleaks/GlobaLeaks#2696]
**Current behavior**
Actually GlobaLeaks do rely on the operating system dns resolver configuration in order to resolve names, this means that in most of the cases, any domain name resolution done by GlobaLeaks happen in a insecure manner over unencrypted, unauthenticated dns query on port 53/udp.

**Expected behavior**
This ticket is to integrate support for DNS over HTTPS configuration for any query made as outgoing connections from GlobaLeaks application code.

By default the DoH servers must be the one used by Mozilla and Google, but it should be possible to change it.

By default this settings should be enabled as a "secure default", but it must be possible to disable DoH, relying on normal operating system name resolution.

In any change of DNS name resolution, before saving the configuration, GlobaLeaks must perform a DNS resolution test with the new settings, confirm on the UX that it's valid and working, and only then save it. In case of problem, the problem should be notified and the new settings not applied.

**Comment by evilaliv3** [comment]
This is actually an interesting topic that recalls me back to the initial globaleaks implementations where we wanted to sandbox every connection and get sure the connection passes from globaleaks.

It would be pretty easy to get sure resolving happens only via the the choosen DoH proxy but it would be actually difficult to get sure all the network operations would pass through it.

**Comment by fpietrosanti** [comment]
We only have outgoing connection in specific point of the code, so we don't need sandboxing, as the code-flow should just know at application level whenever to do name resolution in a way or another:
- Send email
- Check for updates of GlobaLeaks
- Check for Tor Exit List IPs

Are there any point in the code-flow of GlobaLeaks that does outbound connections, so do need to do name resolution?

**Comment by evilaliv3** [comment]
You are correct in the top level interface but most of the time we do not control the internal.

As far as i know most of the time we do not control this name resolutions that are performed transparency below python.

**Comment by fpietrosanti** [comment]
To override the DNS name resolution, specifically for urllib:
https://stackoverflow.com/questions/44374215/how-do-i-specify-url-resolution-in-pythons-requests-library-in-a-similar-fashio
https://stackoverflow.com/questions/22609385/python-requests-library-define-specific-dns

**Comment by fpietrosanti** [comment]
Interesting low-level Twisted DNS resolver hacking https://github.com/scrapy/scrapy/issues/3528#issuecomment-450344010

**Comment by evilaliv3** [comment]
Very interesting code. Will try to see if we could mock once and  It ready to go

**Comment by fpietrosanti** [comment]
While eventually implementing an utility function to achieve that goal, please consider to address also those DNS related issues:
#657 
#2289

**Comment by fpietrosanti** [comment]
Very modern version of Twisted do support Pluggable Name Resolver, rather than mocking it:
https://twistedmatrix.com/documents/18.9.0/api/twisted.internet.interfaces.IReactorPluggableNameResolver.html#installNameResolver
